sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client.py

```diff
@@ -236,16 +236,16 @@ def _create_pipeline(self, credential, **kwargs):
             config.transport = RequestsTransport(**kwargs)
         policies = [
             QueueMessagePolicy(),
-            config.headers_policy,
             config.proxy_policy,
             config.user_agent_policy,
             StorageContentValidation(),
-            StorageRequestHook(**kwargs),
-            self._credential_policy,
             ContentDecodePolicy(response_encoding="utf-8"),
             RedirectPolicy(**kwargs),
             StorageHosts(hosts=self._hosts, **kwargs),
             config.retry_policy,
+            config.headers_policy,
+            StorageRequestHook(**kwargs),
```

Comment:
is it safe to move this after retry policy ? This yields quite many results https://github.com/Azure/azure-sdk-for-python/search?p=1&q=StorageRequestHook. also, is it something customer would use?
if, so putting that after retry might change behavior for them.

@xiafu-msft is this a safe change?

Reply:
I couldn't see a case where this would somehow impact the expected behavior negatively. Having access to the request headers are one of the biggest use cases for using request callbacks for the user. Therefore, the reason this moved after headers was because an expected behavior is being able to access the headers before the request is sent out. for example a user's callback method would be:
```
def _callback(request):
            header = request.http_request.headers.get('x-header')
            if header == "original_value":
                    request.http_request.headers['x-header'] = "changed_value"
```

In this case if the request hook is before the headers policy this would always fail even though the header might be in fact set to `'original_value'`.
If we decide to move the request hook back then that means we would have to move the headers policy back. If we move the headers policy back then the outdated `x-ms-date` header issue would be back. 

So one of two solutions is possible here:
1. Leave the changes made in this PR (i.e, `StorageRequestHook` moved after the new `headers_policy` location)
2. Revert the changes of the headers policy and the storage request hook policy, but remove the `x-ms-date` header setting to be set in a policy after the retry policy (which would mean a non-header policy doing a header operation - which is not great)

Reply:
makes sense. thanks for context.